### PR TITLE
Clean up and improve firewall rules

### PR DESCRIPTION
**Pull request by red2k18**
- [x] Added functionality to delete duplicate firewall rules
- [x] Created service account rule to allow users to SSH into client
- [x] Created service account rule to deny all ingress which overrides the implied rule for deny
- [x] Refactored Installer code

### CODE REVIEW

File: install/gcp/installer/forseti_server_installer.py [hunk 1]
```diff
@@ -36,6 +35,7 @@ class ForsetiServerInstaller(ForsetiInstaller):
     access_target = None
     target_id = None
     user_can_grant_roles = True
+    firewall_rules_to_be_deleted = ["default-allow-icmp", "default-allow-internal", "default-allow-rdp", "default-allow-ssh"]
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: This exceeds the 80-char limiit.

You can do:
```
firewall_rules_to_be_deleted = [
    'default-allow-icmp',
    'default-allow-internal',
    'default-allow-rdp',
    'default-allow-ssh',
]
```

Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: Is `ForsetiServerInstaller` the best place to maintain these firewalls? Since these are not specific to the server. How about moving them to the `ForsetiInstaller` class?

File: install/gcp/installer/forseti_server_installer.py [hunk 2]
```diff
@@ -145,6 +148,28 @@ def create_firewall_rules(self):
             self.config.vpc_host_network,
             '0.0.0.0/0')
 
+        # Create firewall rule to open only port tcp:22 (ssh)
+        # to all the external traffics from the internet to ssh into client VM.
+        gcloud.create_firewall_rule(
+            self.format_firewall_rule_name(
+                'forseti-client-allow-ssh-external'),
+            [self.gcp_service_acct_email],
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: This looks like the server's service account, not the client.

Also, should this be somewhere else?  What about the client installer?  How do you think?

File: install/gcp/installer/forseti_server_installer.py (hunk 1 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: Done

Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: Was thinking the same. Moved it to `ForsetiInstaller` class

File: install/gcp/installer/forseti_server_installer.py [hunk 3]
```diff
@@ -24,7 +24,6 @@
 from util import gcloud
 from util import utils
```
[the remark this answers is not in the document]
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: Ok

File: install/gcp/installer/forseti_client_installer.py [hunk 4]
```diff
@@ -94,3 +96,38 @@ def get_deployment_values(self):
...
+    def format_firewall_rule_name(self, rule_name):
+        """Format firewall rule name.
+
+        Args:
+            rule_name (str): Name of the firewall rule.
+
+        Returns:
+            str: Firewall rule name.
+        """
+        return '{}-{}'.format(rule_name, self.config.identifier)
```
Comment (blueandgold) [inline install/gcp/installer/forseti_client_installer.py]: This is repeated from the method in the server installer, and should be refactored.

File: install/gcp/installer/forseti_server_installer.py [hunk 5]
```diff
@@ -145,17 +153,15 @@ def create_firewall_rules(self):
...
-        return '{}-{}'.format(rule_name, self.config.identifier)
+    def delete_firewall_rules(self):
+        """Deletes default firewall rules as the forseti service account rules
+        serves the purpose"""
+        if len(self.firewall_rules_to_be_deleted) > 0:
+            for rule in self.firewall_rules_to_be_deleted:
+                gcloud.delete_firewall_rule(rule)
+            print('Firewall rules were successfully deleted')
+        else:
+            print('Couldn\'t find rules to delete')
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: This message would be better as: `No firewall rules to delete.`  Thinking on this further, do we really need to display this to the user?

File: install/gcp/installer/forseti_server_installer.py [hunk 6]
```diff
@@ -145,17 +153,15 @@ def create_firewall_rules(self):
...
-            str: Firewall rule name.
-        """
-        return '{}-{}'.format(rule_name, self.config.identifier)
+    def delete_firewall_rules(self):
+        """Deletes default firewall rules as the forseti service account rules
+        serves the purpose"""
+        if len(self.firewall_rules_to_be_deleted) > 0:
+            for rule in self.firewall_rules_to_be_deleted:
+                gcloud.delete_firewall_rule(rule)
+            print('Firewall rules were successfully deleted')
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: I think that we should include what firewall rules where deleted for the users to see.

File: install/gcp/installer/forseti_server_installer.py [hunk 7]
```diff
@@ -101,16 +106,19 @@ def deploy(self, deployment_tpl_path, conf_file_path, bucket_name):
             instance_name = 'forseti-{}-vm-{}'.format(
                 self.config.installation_type,
                 self.config.identifier)
-            self.wait_until_vm_initialized(instance_name)
 
             # Create firewall rules.
             self.create_firewall_rules()
+            self.wait_until_vm_initialized(instance_name)
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: Why is this moved?

File: install/gcp/installer/forseti_server_installer.py [hunk 8]
```diff
@@ -101,16 +106,19 @@ def deploy(self, deployment_tpl_path, conf_file_path, bucket_name):
...
+
+            # Delete firewall rules.
+            self.delete_firewall_rules()
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: These default rules are not server specific, and should be moved outside of the server installer.

File: install/gcp/installer/forseti_client_installer.py [hunk 9]
```diff
@@ -94,3 +96,38 @@ def get_deployment_values(self):
...
+                'forseti-client-allow-ssh-external'),
+            [self.gcp_service_acct_email],
+            constants.FirewallRuleAction.ALLOW,
+            ['tcp:22'],
+            constants.FirewallRuleDirection.INGRESS,
+            0,
+            self.config.vpc_host_network,
+            '0.0.0.0/0')
+
+        # Rule to block out all the ingress traffic on client VM.
```
[the remark this answers is not in the document]
Comment (red2k18) [reply install/gcp/installer/forseti_client_installer.py]: Done

File: install/gcp/installer/forseti_server_installer.py (hunk 5 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: I think we should let the user know whether we deleted it or not.

File: install/gcp/installer/forseti_server_installer.py (hunk 6 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: Consider it done.

File: install/gcp/installer/forseti_server_installer.py (hunk 8 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: It gives VM initialization error if logic to delete default rules is in forseti_installer.py because the rules will be deleted before the server and the client rules are created.

File: install/gcp/installer/forseti_server_installer.py (hunk 7 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: The service account rule to ssh into the VM needs to be created so VM initialization is successful.

File: install/gcp/installer/forseti_client_installer.py (hunk 4 above)
Comment (red2k18) [reply install/gcp/installer/forseti_client_installer.py]: Server installer doesn't have create or delete functionality now. Can be refactored in the future though.

File: install/gcp/installer/forseti_server_installer.py (hunk 5 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: On second thoughts, I deleted the print statement in the else block as it is understood that the rules weren't deleted if they don't see the deleted message.

File: install/gcp/installer/forseti_client_installer.py [hunk 10]
```diff
@@ -15,8 +15,8 @@
 """Forseti CLI installer."""
 
 from forseti_installer import ForsetiInstaller
-
 from util import gcloud
+from util import constants
```
Comment (blueandgold) [inline install/gcp/installer/forseti_client_installer.py]: alphasort this import

File: install/gcp/installer/forseti_server_installer.py (hunk 7 above)
Comment (blueandgold) [reply install/gcp/installer/forseti_server_installer.py]: Ah, that makes sense.  Great move!

File: install/gcp/installer/forseti_server_installer.py (hunk 8 above)
Comment (blueandgold) [reply install/gcp/installer/forseti_server_installer.py]: Hmmm, okay.  I suppose we can can say that the correct firewalls are created only if the server is updated.

Otherwise, the relevant place to do this more centrally would be here.
https://github.com/GoogleCloudPlatform/forseti-security/blob/dev/install/gcp/gcp_installer.py#L122-L137

File: install/gcp/installer/forseti_server_installer.py [hunk 11]
```diff
@@ -101,16 +106,22 @@ def deploy(self, deployment_tpl_path, conf_file_path, bucket_name):
...
 
         return success, deployment_name
 
     def create_firewall_rules(self):
         """Create firewall rules for Forseti server instance."""
-        # Rule to block out all the ingress traffic.
+        # Rule to block out all the ingress traffic on server VM.
+        # We need a service account deny for ingress that is more
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: How does this rephrase look to you?
```
        # This rule overrides the implied deny for ingress
        # because we want a rule with a higher priority
        # that would be harder to be overridden.
```

Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: We need to mention the second reason as well, changed to below
# This rule overrides the implied deny for ingress
# because it has higher priority and is specific to service account.

Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: We need to mention the second reason as well, changed to below
   This rule overrides the implied deny for ingress
   because it has higher priority and is specific to service account.

File: install/gcp/installer/forseti_client_installer.py (hunk 10 above)
Comment (red2k18) [reply install/gcp/installer/forseti_client_installer.py]: Consider it done.

File: install/gcp/installer/forseti_server_installer.py (hunk 8 above)
Comment (red2k18) [reply install/gcp/installer/forseti_server_installer.py]: Based on the flow of control, I have put the logic in server as all the service account rules has to be created before default rules are deleted.

File: install/gcp/installer/forseti_client_installer.py (hunk 4 above)
Comment (blueandgold) [reply install/gcp/installer/forseti_client_installer.py]: Okay

File: install/gcp/installer/forseti_server_installer.py (hunk 8 above)
Comment (blueandgold) [reply install/gcp/installer/forseti_server_installer.py]: Okay

File: install/gcp/installer/forseti_server_installer.py [hunk 12]
```diff
@@ -101,16 +106,20 @@ def deploy(self, deployment_tpl_path, conf_file_path, bucket_name):
...
+            # Delete firewall rules.
+            self.delete_firewall_rules()
 
         return success, deployment_name
 
     def create_firewall_rules(self):
         """Create firewall rules for Forseti server instance."""
-        # Rule to block out all the ingress traffic.
+        # This rule overrides the implied deny for ingress
+        # because it has higher priority and is specific to service account.
```
Comment (blueandgold) [inline install/gcp/installer/forseti_server_installer.py]: I want to add to this comment, the "why", which is to make it harder to override.  Can you please add this?  Here and in the client?

File: install/gcp/installer/forseti_client_installer.py [hunk 13]
```diff
@@ -94,3 +97,40 @@ def get_deployment_values(self):
...
+        # Create firewall rule to open only port tcp:22 (ssh)
+        # to all the external traffic from the internet to ssh into client VM.
+        gcloud.create_firewall_rule(
+            self.format_firewall_rule_name(
+                'forseti-client-allow-ssh-external'),
+            [self.gcp_service_acct_email],
+            constants.FirewallRuleAction.ALLOW,
+            ['tcp:22'],
+            constants.FirewallRuleDirection.INGRESS,
+            0,
```
[the remark this answers is not in the document]
Comment (red2k18) [reply install/gcp/installer/forseti_client_installer.py]: Done

File: install/gcp/installer/forseti_client_installer.py [hunk 14]
```diff
@@ -94,3 +97,40 @@ def get_deployment_values(self):
...
+            '0.0.0.0/0')
+
+        # This rule overrides the implied deny for ingress
+        # because it is specific to service account with a higher priority
+        # that would be harder to be overriden.
+        gcloud.create_firewall_rule(
+            self.format_firewall_rule_name('forseti-client-deny-all'),
+            [self.gcp_service_acct_email],
+            constants.FirewallRuleAction.DENY,
+            ['icmp', 'udp', 'tcp'],
```
[the remark this answers is not in the document]
Comment (red2k18) [reply install/gcp/installer/forseti_client_installer.py]: Done
